src/ripple/app/main/GRPCServer.h

```diff
@@ -0,0 +1,349 @@
+                Handler<Request,Response> handler,
+                RPC::Condition required_condition,
+                Resource::Charge load_type);
+   
+  
+        void process() override;
+  
+        bool isFinished() override
+        {
+            //checking the status while a request is in the middle of being
+            //processed will lead to indeterminate results. Lock here to 
+            //sequence checking status and processing
+            std::lock_guard<std::mutex> lock(mut_);
+            return status_ == CallStatus::FINISH;
+        }
+  
+        virtual void abort() override
+        {
+            std::lock_guard<std::mutex> lock(mut_);
+            aborted_ = true;
+        }
+  
+        void set_iter(std::list<std::shared_ptr<Processor>>::iterator const & it) override
+        {
+            iter_ = it;
+        }
+  
+        std::list<std::shared_ptr<Processor>>::iterator get_iter() override
+        {
+            if(!iter_)
+                BOOST_ASSERT(false);
+  
+            return iter_.get();
+        }
+  
+        std::shared_ptr<Processor> clone() override;
+  
+        private:
+  
+        //process the request. Called inside the coroutine passed to JobQueue
+        void process(std::shared_ptr<JobQueue::Coro> coro);
+  
+        //return load type of this RPC
+        Resource::Charge getLoadType()
+        {
+            return load_type_;
+        }
+  
+        //return the Role required for this RPC
+        //for now, we are only supporting RPC's that require Role::USER for gRPC
+        Role getRole()
+        {
+            return Role::USER;
+        }
+  
+        //register endpoint with ResourceManager and return usage
+        Resource::Consumer getUsage()
+        {
+            std::string peer = getEndpoint(ctx_.peer());
+            boost::optional<beast::IP::Endpoint> endpoint =
+                beast::IP::Endpoint::from_string_checked(peer);
+            return app_.getResourceManager().newInboundEndpoint(endpoint.get());
+        }
+
+    }; //CallData
+  
+  
+    //Create a CallData object for each RPC
+    void setupListeners();
+  
+    //make a CallData instance, returned as shared_ptr to base class (Processor)
+    template <class Request, class Response>
+    std::shared_ptr<Processor> makeCallData(
+            BindListener<Request,Response> bl,
+            Handler<Request,Response> handler,
+            RPC::Condition condition,
+            Resource::Charge load_type)
+    {
+        auto ptr = std::make_shared<CallData<Request,Response>>(
+                service_,*cq_,app_,bl,handler, condition, load_type);
+        return std::static_pointer_cast<Processor>(ptr);
+    }
+  
+    //make CallData instance and push to requests list
+    template <class Request,class Response>
+    void makeAndPush(
+            BindListener<Request,Response> bl,
+            Handler<Request,Response> handler,
+            RPC::Condition condition,
+            Resource::Charge load_type)
```

Comment:
If we're getting rid of storing the iterator with the object, I'd remove this function.

Reply:
Will remove this function as well

Reply:
@seelabs I actually don't think this function should be removed. While we don't need to set the iterator, we still need to add each `shared_ptr<CallData>` object to the `requests_` collection. Without this function, I would have to write `requests_.push_front()` (or whatever insertion method is used) for each RPC, as opposed to just calling this function

Reply:
We have to call makeAndPush for each RPC now. The difference is we replace code like
```
makeAndPush<rpc::v1::GetFeeRequest, rpc::v1::GetFeeResponse>(
            &rpc::v1::XRPLedgerAPIService::AsyncService::RequestGetFee,
            doFeeGrpc,
            RPC::NEEDS_CURRENT_LEDGER,
            Resource::feeReferenceRPC
            );
```
with
```
{
        using cd = CallData<rpc::v1::GetFeeRequest, rpc::v1::GetFeeResponse>;
        requests_.push_front(std::make_shared<cd>(
            service_, *cq_, app_,
            &rpc::v1::XRPLedgerAPIService::AsyncService::RequestGetFee,
            doFeeGrpc, RPC::NEEDS_CURRENT_LEDGER, Resource::feeReferenceRPC));
}
```
You do save repeating the ` service_, *cq_, app_,` parameters with a helper function, but I'm not sure it's worth defining a helper function just for that. I don't really object if you want to keep a helper function like this, there's not great harm, but I also don't see much benefit either.

Reply:
removed